File: CS/E3802/HotTrackHelper.cs
using System;
using System.Collections.Generic;

using System.Text;
using DevExpress.XtraTreeList;
using DevExpress.XtraTreeList.Nodes;
using DevExpress.XtraEditors;
using DevExpress.XtraTreeList.ViewInfo;
using System.Drawing;
using DevExpress.XtraEditors.ViewInfo;
using DevExpress.XtraEditors.Controls;
using DevExpress.XtraEditors.Drawing;

namespace treelist
{
    class HotTrackHelper
    {
        TreeList fTreeList;
        TreeListNode hotNode;
        SkinEditorButtonPainter customButtonPainter;
        EditorButtonObjectInfoArgs args;

        public HotTrackHelper(TreeList treeList)
        {
            Attach(treeList);
        }
        public void Attach(TreeList treeList)
        {
            fTreeList = treeList;
            fTreeList.CustomDrawNodeIndent += new CustomDrawNodeIndentEventHandler(Tree_CustomDrawNodeIndent);
            fTreeList.CustomDrawNodeCell += new CustomDrawNodeCellEventHandler(Tree_CustomDrawNodeCell);
            fTreeList.MouseMove += new System.Windows.Forms.MouseEventHandler(Tree_MouseMove);

            fTreeList.OptionsSelection.EnableAppearanceFocusedCell = false;
            CreateButtonInfoArgs();
            CreateButtonPainter();
        }

        private void CreateButtonInfoArgs() {
            EditorButton btn = new EditorButton(ButtonPredefines.Glyph);
            
            args = new EditorButtonObjectInfoArgs(btn, new DevExpress.Utils.AppearanceObject());
            args.Transparent = true;
        }
        private void CreateButtonPainter() {
            customButtonPainter = new SkinEditorButtonPainter(DevExpress.LookAndFeel.UserLookAndFeel.Default.ActiveLookAndFeel);
        }
        void Tree_MouseMove(object sender, System.Windows.Forms.MouseEventArgs e)
        {
            TreeListHitTest hi = fTreeList.ViewInfo.GetHitTest(e.Location);
            {
                if (hotNode != hi.Node)
                {
                    fTreeList.RefreshNode(hotNode);
                    hotNode = hi.Node;
                    fTreeList.RefreshNode(hotNode);
                }
            }
        }
        void Tree_CustomDrawNodeCell(object sender, CustomDrawNodeCellEventArgs e)
        {
            DrawBackground(e.Node, e);
        }
        void Tree_CustomDrawNodeIndent(object sender, CustomDrawNodeIndentEventArgs e)
        {
            DrawBackground(e.Node, e);
            e.Handled = true;
        }
        void DrawBackground(TreeListNode node, CustomDrawEventArgs e)
        {
            if (hotNode != node && !node.Focused) return;
            SetUpButtonInfoArgs(e);
            customButtonPainter.DrawObject(args);
        }
        private void SetUpButtonInfoArgs(CustomDrawEventArgs e) {
            args.Cache = e.Cache;
            args.Bounds = e.Bounds;
            args.State = DevExpress.Utils.Drawing.ObjectState.Hot;
        }
        public void Detach()
        {
            fTreeList.CustomDrawNodeIndent -= new CustomDrawNodeIndentEventHandler(Tree_CustomDrawNodeIndent);
            fTreeList.CustomDrawNodeCell -= new CustomDrawNodeCellEventHandler(Tree_CustomDrawNodeCell);
            fTreeList.MouseMove -= new System.Windows.Forms.MouseEventHandler(Tree_MouseMove);
            fTreeList.OptionsSelection.EnableAppearanceFocusedCell = true;
            fTreeList = null;
        }
    }
}

File: VB/E3802/HotTrackHelper.vb
Imports System
Imports System.Collections.Generic

Imports System.Text
Imports DevExpress.XtraTreeList
Imports DevExpress.XtraTreeList.Nodes
Imports DevExpress.XtraEditors
Imports DevExpress.XtraTreeList.ViewInfo
Imports System.Drawing
Imports DevExpress.XtraEditors.ViewInfo
Imports DevExpress.XtraEditors.Controls
Imports DevExpress.XtraEditors.Drawing

Namespace treelist
    Friend Class HotTrackHelper
        Private fTreeList As DevExpress.XtraTreeList.TreeList
        Private hotNode As TreeListNode
        Private customButtonPainter As SkinEditorButtonPainter
        Private args As EditorButtonObjectInfoArgs

        Public Sub New(ByVal treeList As DevExpress.XtraTreeList.TreeList)
            Attach(treeList)
        End Sub
        Public Sub Attach(ByVal treeList As DevExpress.XtraTreeList.TreeList)
            fTreeList = treeList
            AddHandler fTreeList.CustomDrawNodeIndent, AddressOf Tree_CustomDrawNodeIndent
            AddHandler fTreeList.CustomDrawNodeCell, AddressOf Tree_CustomDrawNodeCell
            AddHandler fTreeList.MouseMove, AddressOf Tree_MouseMove

            fTreeList.OptionsSelection.EnableAppearanceFocusedCell = False
            CreateButtonInfoArgs()
            CreateButtonPainter()
        End Sub

        Private Sub CreateButtonInfoArgs()
            Dim btn As New EditorButton(ButtonPredefines.Glyph)

            args = New EditorButtonObjectInfoArgs(btn, New DevExpress.Utils.AppearanceObject())
            args.Transparent = True
        End Sub
        Private Sub CreateButtonPainter()
            customButtonPainter = New SkinEditorButtonPainter(DevExpress.LookAndFeel.UserLookAndFeel.Default.ActiveLookAndFeel)
        End Sub
        Private Sub Tree_MouseMove(ByVal sender As Object, ByVal e As System.Windows.Forms.MouseEventArgs)
            Dim hi As TreeListHitTest = fTreeList.ViewInfo.GetHitTest(e.Location)
            If True Then
                If hotNode IsNot hi.Node Then
                    fTreeList.RefreshNode(hotNode)
                    hotNode = hi.Node
                    fTreeList.RefreshNode(hotNode)
                End If
            End If
        End Sub
        Private Sub Tree_CustomDrawNodeCell(ByVal sender As Object, ByVal e As CustomDrawNodeCellEventArgs)
            DrawBackground(e.Node, e)
        End Sub
        Private Sub Tree_CustomDrawNodeIndent(ByVal sender As Object, ByVal e As CustomDrawNodeIndentEventArgs)
            DrawBackground(e.Node, e)
            e.Handled = True
        End Sub
        Private Sub DrawBackground(ByVal node As TreeListNode, ByVal e As CustomDrawEventArgs)
            If hotNode IsNot node AndAlso Not node.Focused Then
                Return
            End If
            SetUpButtonInfoArgs(e)
            customButtonPainter.DrawObject(args)
        End Sub
        Private Sub SetUpButtonInfoArgs(ByVal e As CustomDrawEventArgs)
            args.Cache = e.Cache
            args.Bounds = e.Bounds
            args.State = DevExpress.Utils.Drawing.ObjectState.Hot
        End Sub
        Public Sub Detach()
            RemoveHandler fTreeList.CustomDrawNodeIndent, AddressOf Tree_CustomDrawNodeIndent
            RemoveHandler fTreeList.CustomDrawNodeCell, AddressOf Tree_CustomDrawNodeCell
            RemoveHandler fTreeList.MouseMove, AddressOf Tree_MouseMove
            fTreeList.OptionsSelection.EnableAppearanceFocusedCell = True
            fTreeList = Nothing
        End Sub
    End Class
End Namespace
